Windows Support

yodaldevoid opened the pull request:
The template uses Windows Server Core as the base image for now, though
really any Windows image could be a base. The Dockerfile could almost
definitely be optimized, but this works for now. All Windows versions
that the base image supports are generated by x.py.

TODO before merging:
[ ] Appveyor (or some other CI system) support for updating the image
[ ] Maybe support the GNU toolchains?

sfackler commented on Dockerfile-windows.template:
```diff
@@ -0,0 +1,31 @@
+ENV RUST_VERSION=%%RUST-VERSION%%
+
+# Install the Visual Studio Build Tools.
+ADD https://aka.ms/vs/16/release/vs_buildtools.exe /vs_buildtools.exe
+RUN vs_buildtools.exe --quiet --wait --norestart --nocache \
+        --installPath C:\BuildTools \
+        --add Microsoft.Component.MSBuild \
+        --add Microsoft.VisualStudio.Component.Windows10SDK.%%SDK-BUILD%% \
+        --add Microsoft.VisualStudio.Component.VC.Tools.x86.x64 \
+    || IF "%ERRORLEVEL%"=="3010" EXIT 0
+RUN del vs_buildtools.exe
+RUN setx /M __VSCMD_ARG_NO_LOGO "1"
+
+ENTRYPOINT ["powershell.exe", "-NoLogo", "-ExecutionPolicy", "Bypass"]
```
This block of steps should be merged into a single one to keep the total image size down.

sfackler commented on Dockerfile-windows.template:
```diff
@@ -0,0 +1,31 @@
+ENTRYPOINT ["powershell.exe", "-NoLogo", "-ExecutionPolicy", "Bypass"]
+
+# Install Rustup and Rust toolchains for x86 and x64
+ADD https://static.rust-lang.org/rustup/archive/%%RUSTUP-VERSION%%/x86_64-pc-windows-msvc/rustup-init.exe /rustup-init.exe
+RUN rustup-init.exe -y -v --default-toolchain %RUST_VERSION% --default-host x86_64-pc-windows-msvc
+RUN del rustup-init.exe
+RUN rustup toolchain install %RUST_VERSION%-i686-pc-windows-msvc
```
Same here

sfackler commented on x.py:
```diff
@@ -38,6 +38,16 @@
 default_alpine_version = "3.12"
 
+windows_versions = [
+    ("1709", "16299"),
+    ("1803", "17134"),
+    ("1809", "17763"),
+    ("1903", "18362"),
+    ("1909", "18362"),
+    ("2004", "19041"),
+    ("20H2", "19041"),
```
Do other official images cover this many Windows revisions?

yodaldevoid replied:
This list includes all of the non-duplicated Windows revisions that are in the full tag listing of the [Windows Server Core Docker Hub page](https://hub.docker.com/_/microsoft-windows-servercore).

sfackler replied:
It looks like golang's currently just provide 1809 and ltsc2016: https://hub.docker.com/_/golang

yodaldevoid replied:
Limiting to just 1809 is easy enough, but I don't think we can support 1607/ltsc2016 as the Visual Studio installer doesn't have a matching/compatible Windows 10 SDK version available. We can only support up to 1809 using Github Actions or Travis because that is the latest version of Windows Server they support. I have not looked into Azure's support.

sfackler replied:
Okay, let's just go with 1809 for now.

sfackler commented on Dockerfile-windows.template:
```diff
@@ -0,0 +1,31 @@
+ADD https://static.rust-lang.org/rustup/archive/%%RUSTUP-VERSION%%/x86_64-pc-windows-msvc/rustup-init.exe /rustup-init.exe
+RUN rustup-init.exe -y -v --default-toolchain %RUST_VERSION% --default-host x86_64-pc-windows-msvc
+RUN del rustup-init.exe
+RUN rustup toolchain install %RUST_VERSION%-i686-pc-windows-msvc
```
Why install the 32 bit toolchain as well by default?

yodaldevoid replied:
My original use case requires both the 32 and 64 bit versions of the Rust toolchain for testing, so I installed both. If you would like, I can remove the 32 bit toolchain.

sfackler replied:
Yeah I think we should leave it to just x86_64 in the base image. It's a quick command away if you need another target

sfackler commented:
On the CI side, I think github actions probably makes sense, and we should probably switch the Linux builds over from Travis while we're at it.

yodaldevoid commented:
Shoutout to @nelsonjchen for his previous work on this. It helped a lot when minimizing the dockerfiles.

I've minimized everything as far as possible without breaking something. I've added in this PR dockerfiles for the MSVC and GNU toolchains. Also, I've moved all of the CI builds over to GitHub Actions. It is crazy how easy it is to reuse the entire testing framework is between all of the versions built.

sfackler commented on 1.50.0/windowsservercore-1809/gnu/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+FROM mcr.microsoft.com/windows/servercore:1809
+
+SHELL ["powershell.exe", "-NoLogo", "-ExecutionPolicy", "Bypass", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]
```
It looks like golang's image doesn't set `-NoLogo` or `-ExecutionPolicy Bypass`: https://github.com/docker-library/golang/blob/master/1.15/windows/windowsservercore-1809/Dockerfile#L10. Do you know why there's a difference there?

yodaldevoid replied:
I set `-NoLogo` because I don't see a reason to have the PowerShell version information and copyright info printed at the start of each command as it's just noise.

As for the `-ExecutionPolicy Bypass`, I think that might be a leftover from another Windows docker tutorial I had followed. I don't see a reason it needs to be kept.

sfackler commented on 1.50.0/windowsservercore-1809/msvc/Dockerfile:
```diff
@@ -0,0 +1,72 @@
+# escape=`
+
+# IMAGES BUILT FROM THIS DOCKERFILE ARE NOT TO BE SHARED ON PUBLIC DOCKER HUBs.
```
The intent of this PR is to produce images that are to be shared on public Docker hubs, right?

yodaldevoid replied:
I'm not entirely sure at this point. I had completely missed this license issue until I was reading through @nelsonjchen's implementation. This issue was brought up by them before a few years ago and I didn't see any discussion about it.

I see value in providing a dockerfile for a base Rust on Windows image, but I don't know whether this repo is the right place for that. If we do keep this base dockerfile here, we would just need to make sure not to add it to the manifest file. (Which reminds me, I need to modify x.py to add the GNU builds to the manifest file...)

nelsonjchen commented:
Could you put "`Co-authored-by: Nelson Chen <crazysim@gmail.com>`" at the end of the commit messages where I contributed to the Dockerfiles?

lippertmarkus commented:
How can we bring this forward?

yodaldevoid commented:
@lippertmarkus I think the one open question is if we want to have the MSVC toolchain dockerfile/template in this repo despite the fact that it can currently never be uploaded to Docker Hub.

I could split this PR up so that all changes but the MSVC toolchain can be merged now.

yodaldevoid commented:
@lippertmarkus I will ask, but going off of paragraph four of [this official blog post](https://devblogs.microsoft.com/cppblog/using-msvc-in-a-docker-container-for-your-c-projects/) (which is called out in the big wall of text in the MSVC Dockerfile), any Dockerfiles with the MSVC build tools are only meant for personal use or use within a organization. They even explicitly note that you should not upload containers containing the MSVC build tools to a public Docker hub.

lippertmarkus commented:
Thanks for sharing. That's interesting as Microsoft is using the Build Tools in their official dotnet framework docker images (https://github.com/microsoft/dotnet-framework-docker/blob/3b2ac5bcda7603b94c0b6399355f3d0c0c1eb512/src/sdk/4.8/windowsservercore-1909/Dockerfile#L21-L30) and that would mean you wouldn't be allowed to use those official images without having a Visual Studio License 😄 

Let's see what they're saying.

lippertmarkus commented:
I created https://github.com/microsoft/Windows-Containers/issues/102

yodaldevoid commented:
> I created [microsoft/Windows-Containers#102](https://github.com/microsoft/Windows-Containers/issues/102)

Thank you for that. My free time suddenly disappeared last week and this feel by the wayside.

sfackler commented:
Hey, sorry for dropping this for a while!

The main concern is around the licensing of the VS components as you've been discussing upthread. It seems a bit unfortunate to only publish a GNU-based image as MSVC has been the recommended Windows target for quite a while.

lippertmarkus commented:
seems like we can't continue this work (see https://github.com/microsoft/Windows-Containers/issues/102) 😞

yodaldevoid commented:
> 
> 
> seems like we can't continue this work (see [microsoft/Windows-Containers#102](https://github.com/microsoft/Windows-Containers/issues/102)) 😞

Well, to be specific, we can't upload any Docker image created from the MSVC Dockerfile to Docker Hub. We could keep an example template in this or another repo, which I think has a lot of value. All the other changes made by this PR should be fine, however.

lippertmarkus commented:
Without any official image to build rust binaries, I'm now cross-building my Windows container images on Linux instead: https://github.com/rust-lang/docker-rust/issues/30#issuecomment-985261684 instead

haxtibal commented:
To my understanding, solving rust-lang/rust#71520 would reopen the door for this PR here.  

This PR seems stalled because it [depends on vs_buildtools.exe](https://github.com/rust-lang/docker-rust/blob/ba375b78b82507ebae9242b294f3183a7cb8c22d/1.50.0/windowsservercore-1809/msvc/Dockerfile#L39) and [redistribution of the VS Build Tools is not permitted](https://github.com/microsoft/Windows-Containers/issues/102#issuecomment-827170844). In turn, VS Buildtools are required because Rust uses their `link.exe` as default linker.  With rust-lang/rust#71520, VS Buildtools would no longer be required. Instead, Rusts copy of `lld` would be used. Such an image would be good for public upload.

In most situations it's possible to use [lld](https://lld.llvm.org/) for `x86_64-pc-windows-msvc` by now.  We could do some preliminary trials by setting
```toml
[target.x86_64-pc-windows-msvc]
linker = "rust-lld"
```
or `CARGO_TARGET_X86_64_PC_WINDOWS_MSVC_LINKER="rust-lld"`.

What do you think?

ChrisDenton commented:
See also [this internals thread](https://internals.rust-lang.org/t/pre-rfc-remove-rusts-dependency-on-visual-studio-in-4-complex-steps/16708/12) for an overview. In short, vcruntime may one day be made open source, which would help solve issues such as this, but that day isn't today.

yodaldevoid commented:
> In short, vcruntime may one day be made open source, which would help solve issues such as this, but that day isn't today.

It looks like the Microsoft team that was pushing for this has abandoned the effort. See the paragraph at the bottom of [this comment](https://github.com/microsoft/STL/issues/4560#issuecomment-2348622212).

bruingineer commented:
I forked and rebased here: https://github.com/bruingineer/docker-rust/tree/windows
I also switched the gnu image to download "x86_64-pc-windows-gnu" version of rustup and set "x86_64-pc-windows-gnu" as the default host. I added msys2 to the gnu image so gcc can be used and msvc is avoided. 

The update ci funtions on that branch are not caught up with some of the changes I made.

Whether or not this is helpful to the overall rust-docker project, I figured there may be some people interested in this. (maybe even dozens of us!)
